src/persist-txn/src/txns.rs

```diff
@@ -550,6 +550,39 @@ where
         Tidy { retractions }
     }
 
+    /// [Self::apply_le] but also advances the physical upper of every data
+    /// shard registered at the timestamp past the timestamp.
+    #[instrument(level = "debug", skip_all, fields(ts = ?ts))]
+    pub(crate) async fn apply_eager_le(&mut self, ts: &T) -> Tidy {
+        // TODO: Ideally the upper advancements would be done concurrently with
+        // this apply_le.
+        let tidy = self.apply_le(ts).await;
+
```

Comment:
Does anything bad happen if we crash here, after the `apply_le` but before the advancing the physical uppers? One potentially bad scenario:

1. Upper of some shard `s` is at `5`.
2. Apply txns at `10`.
3. Crash.
4. Reboot.
5. Try and read `s` at `10`.

If at some point during startup we call `apply_eager_le(ts')` s.t. `ts' >= 10`, then I think we're fine. Otherwise we might need to wait until the next commit to read the shard?

Reply:
Good point, seems like we just need call `apply_eager_le` in init_ts?

Reply:
That sounds like it would work.

Reply:
Now that I think about it though, we definitely do at least one table write during bootstrap so it's probably fine as is. Might be better to be explicit about it though.

Reply:
Yeah, I realized the same thing and ditto for create_collection, which would also advance them. Added it anyway :), but had to do it in register instead of init_txns.